doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,209 @@
+and `(row, upper_ts, -1)`. The `lower_ts` and `upper_ts` timestamps are entirely dependent on the
+expression the user typed and can be arbitrarily in the future.. This is best depicted in this
+diagram:
+
+![Temporal filter application](https://github.com/MaterializeInc/materialize/raw/9c79faf5d58b3a97b11f982fa4f6e7736c864922/doc/developer/design/static/temporal_filters.png)
+
+This is a problem because when the `RAW` colletion produces the differential update `(row, t, +1)`
+we might have to emit it at some future time `t_temporal >> t` but if we crash and restart we might
+have to resume from some time `t_resume > t && t_resume < t_temporal`. In this case we will entirely
+miss the original update from `RAW` and we will fail to produce the differential update that we
+would have produced had we not crashed.
+
+In order to solve this we will need to decompose `Q_mfp` into three parts: `Q_safemfp`,
+`Q_temporal_filter`, `Q_final_projection`
+
+> 💡 To understand the following decomposition it is important to keep in mind that all temporal
+> predicates are internally converted into expressions that produce a candidate timestamp for either
+> the lower or upper bound of the validity period. For example, the predicate `expr <= mz_now()`
+> will get converted into just the expression `expr` and will be stored in a `lower_bounds` list. To
+> calculate the lower and upper bounds of the validity period of a row during a temporal MFP
+> evaluation all the lower bound expressions are evaluated and the maximum is selected and all the
+> upper bound expressions are evaluated and minimum is selected.
+
+The goal of the decomposition is to come up with a `Q_safemfp` that contains no temporal filters,
+performs the maximum amount of mapping and filtering, and retains enough information to later
+recover the original validity period. We can define the three parts like so:
+
+1. `Q_safemfp` will contain the non-temporal part of `Q_mfp` combined with two additional `Map`
+   expressions. These two additional map expressions will correspond to the lower and upper bounds
+   and will be calculating the final timestamp for them. We can do so by fabricating two
+   `MirScalarExpr`s. One that calculates the maximum of all the lower bound exprs and one that
+   calculates the minimum of all the upper bound exprs. Effectively, we will precompute the validity
+   period and store it as two additional timestamp columns at the end of the original row.
+2. `Q_temporal_filter` will contain two trivial temporal predicates:
+    1. `lower_ts <= mz_now()`
+    2. `mz_now() <= upper_ts`
+3. `Q_final_projection` will be simply throwing away the two last columns (`lower_ts` and
+   `upper_ts`) of the row
+
+Given this decomposition we can now define the ingestion as:
+
+1. `C_intermediate = PERSIST(Q_temporal_filter(Q_safemfp(RAW)))`
+2. `C = PERSIST(Q_expr(Q_final_projection(Q_temporal_filter(C_intermediate))))`
```

Comment:
I'm not sure I follow this, or I missed something. I think we want to apply half of `Q_temporal_filter` in (1.), specifically the upper bound, and half of `Q_temporal_filter` in (2.), specifically the lower bound. The text here seems to imply you just apply them both, twice. But I can't understand (1.) actually applying the lower bound: doesn't this end up in the problematic state where if the record is withheld at all, we risk missing it completely?

Reply:
You're 100% correct! I realized last night that the equations are wrong but was already in bed :D 

I updated the equations and I think the logic is correct now. `Q_mfp` evalues the entirety of the original MFP and promotes the temporal filters into a `lower_ts` and `upper_ts` columns. Then `Q_expiration` only expires records (and leaves the validity columns intact) and that is what we persist as `C_intermediate`. Then on the other side we apply a `Q_final` which only fast-forwards records to their `lower_ts` and hides the two additional columns from `Q_expr`. The text in the doc goes in more detail